mlir/test/Dialect/LLVMIR/constant-folding.mlir

```diff
@@ -101,3 +101,63 @@ llvm.func @addressof_blocks(%arg: i1) -> !llvm.ptr {
 }
 
 llvm.mlir.global constant @foo() : i32
+
+// -----
+
+// CHECK-LABEL: llvm.func @zero
+llvm.func @zero() {
+  // CHECK-NEXT: %[[ZERO:.+]] = llvm.mlir.zero : i32
+  %zero1 = llvm.mlir.zero : i32
+  %zero2 = llvm.mlir.zero : i32
+  // CHECK-NEXT: llvm.call @foo(%[[ZERO]], %[[ZERO]])
+  llvm.call @foo(%zero1, %zero2) : (i32, i32) -> ()
+  // CHECK-NEXT: llvm.return
+  llvm.return
+}
+
+llvm.func @foo(i32, i32)
+
+// -----
+
+// CHECK-LABEL: llvm.func @null_pointer
+llvm.func @null_pointer() {
+  // CHECK-NEXT: %[[NULLPTR:.+]] = llvm.mlir.zero : !llvm.ptr
+  %nullptr1 = llvm.mlir.zero : !llvm.ptr
+  %nullptr2 = llvm.mlir.zero : !llvm.ptr
+  // CHECK-NEXT: llvm.call @foo(%[[NULLPTR]], %[[NULLPTR]])
+  llvm.call @foo(%nullptr1, %nullptr2) : (!llvm.ptr, !llvm.ptr) -> ()
+  // CHECK-NEXT: llvm.return
+  llvm.return
+}
+
+llvm.func @foo(!llvm.ptr, !llvm.ptr)
+
+// -----
+
+// CHECK-LABEL: llvm.func @undef
+llvm.func @undef() {
+  // CHECK-NEXT: %[[UNDEF:.+]] = llvm.mlir.undef : i32
+  %undef1 = llvm.mlir.undef : i32
+  %undef2 = llvm.mlir.undef : i32
+  // CHECK-NEXT: llvm.call @foo(%[[UNDEF]], %[[UNDEF]])
+  llvm.call @foo(%undef1, %undef2) : (i32, i32) -> ()
+  // CHECK-NEXT: llvm.return
+  llvm.return
+}
+
+llvm.func @foo(i32, i32)
+
+// -----
```

Comment:
There are some more tests that are interesting. I think we can test that a switch that gets two zero attributes folds to just zero (undef / poison would probably have special handling that we do not yet implement in LLVM dialect)... Another interesting test could be if we can fold the zext of an zero constant. Folding an addition would be nice as well but LLVM dialect does not implement such foldings...

Reply:
I've added a test for a `select` of null-pointers. Do you think that testing `zext` is also beneficial? It seems a bit like boiler-plate matrix of tests for each constant-like someone wishes to add.

Reply:
If we fold mlir.zero : i32 to an integer attribute as mentioned above then there would be some value since the zero operation should disappear and fold with the zext.